Sources/Plasma/FeatureLib/pfConsole/pfRemoteConsole.cpp

```diff
@@ -0,0 +1,206 @@
+        PyObject* mymod = PythonInterface::FindModule("__main__");
+        PythonInterface::RunStringInteractive(cmd->GetCommand().c_str(), mymod);
+        std::string output;
+        // get the messages
+        PythonInterface::getOutputAndReset(&output);
+        cmd->GetOutput().get()->fOutputData = ST::string::from_std_string(output);
+        {
+            std::lock_guard<std::mutex> lk(cmd->GetOutput().get()->fOutputDataSetLock);
+            cmd->GetOutput().get()->fOutputDataSet = true;
+        }
+        cmd->GetOutput().get()->fCondvar.notify_one();
+
+        return true;
+    }
+    return hsKeyedObject::MsgReceive(msg);
+}
+
+bool canRun(ST::string& script) {
+    return script.contains("\n");
+}
+
+void handlePipeServer() {
+    SetThreadDescription(
+        GetCurrentThread(),
+        L"PipeServer!"
+    );
+
+    // First, initialize the socket
+    DWORD pid = GetCurrentProcessId();
+    ST::string pipename = ST::format(R"(\\.\pipe\URU-PYTHON-{})", pid);
+
+    HANDLE pipe = CreateNamedPipeW(
+        pipename.to_std_wstring().c_str(),
+        PIPE_ACCESS_DUPLEX,
+        PIPE_TYPE_BYTE | PIPE_READMODE_BYTE,
+        1,
+        // 1MiB
+        1 * 1024 * 1024,
+        1 * 1024 * 1024,
+        0,
+        nullptr
+    );
+    if (pipe == nullptr) {
+        // TODO: Log an error.
+        return;
+    }
+
+    if (!ConnectNamedPipe(pipe, nullptr) && GetLastError() != ERROR_PIPE_CONNECTED) {
+        DWORD err = GetLastError();
+        // TODO: Log an error
+        CloseHandle(pipe);
+        return;
+    }
+
+
+    char buf;
+
+    ST::string curScript;
+
+    SendFullBuf(pipe, ">>> ", 4);
+
+    bool isMultiline = false;
+    while (true) {
+        DWORD readSize;
+        if (!ReadFile(pipe, &buf, sizeof(buf), &readSize, nullptr)) {
+            DWORD err = GetLastError();
+            CloseHandle(pipe);
+            // TODO: Log the error.
+            return;
+        }
+
+        if (readSize == 0) {
+            // TODO: Log that pipe is closed.
+            CloseHandle(pipe);
+            return;
+        }
+
+        if (buf != '\r') {
+            SendFullBuf(pipe, &buf, 1);
+        } else {
+            SendFullBuf(pipe, "\r\n", 2);
+        }
+
+        // Turn our string into ST
+        ST::string readData(&buf, readSize);
+
+        readData = readData.replace("\r", "\n");
+        curScript += readData;
+
+        if (buf != '\r') {
+            continue;
+        }
+
+        if (!isMultiline && curScript.ends_with(":\n")) {
+            isMultiline = true;
+            SendFullBuf(pipe, "... ", 4);
+            continue;
+        }
+
+        if (isMultiline && !curScript.ends_with("\n\n")) {
+            SendFullBuf(pipe, "... ", 4);
+            continue;
+        }
+
+        // Send message.
+        pfRemoteConsoleMsg *cMsg = new pfRemoteConsoleMsg(curScript);
+        std::shared_ptr<pfRemoteConsoleMsgOutput> output(cMsg->GetOutput());
+
+        cMsg->Send(nullptr, true);
```

Comment:
Based on the implementation of [`plDispatch::IMsgEnqueue`](https://github.com/H-uru/Plasma/blob/37d26cae22728af72a6abcb9676f102421654488/Sources/Plasma/NucleusLib/pnDispatch/plDispatch.cpp#L231-L250) this should actually be safe from to call from a thread, as long as the second parameter is `true`.

It will lock a mutex before adding it to the queue of messages to be dispatched.

If the second parameter is false, it will also immediately try to dispatch it and probably call the receiver's `MsgReceive` handler on the thread which is probably not safe.